lisa/util/process.py

```diff
@@ -16,86 +25,135 @@ def __init__(self, level: int):
         and start the thread
         """
         Thread.__init__(self)
+        self.output: str = ""
         self.daemon = False
         self.level = level
         self.fdRead, self.fdWrite = os.pipe()
         self.pipeReader = os.fdopen(self.fdRead)
+        self.isReadCompleted = False
+        self.isClosed = False
         self.start()
 
-    def fileno(self):
+    def fileno(self) -> int:
         """Return the write file descriptor of the pipe
         """
         return self.fdWrite
 
-    def run(self):
+    def run(self) -> None:
         """Run the thread, logging everything.
         """
-        for line in iter(self.pipeReader.readline, ""):
-            log.log(self.level, line.strip("\n"))
+        output = self.pipeReader.read()
+        self.output = "".join([self.output, output])
+        for line in output.splitlines(False):
+            log.log(self.level, line)
 
         self.pipeReader.close()
+        self.isReadCompleted = True
 
-    def close(self):
+    def close(self) -> None:
         """Close the write end of the pipe.
         """
-        os.close(self.fdWrite)
+        if not self.isClosed:
+            os.close(self.fdWrite)
+            self.isClosed = True
 
 
 class Process:
```

Comment:
We should definitely do a thorough review of this design and implementation (more than I can do here).

Reply:
Yes, I had a backlog item to handle subprocesses well. It looks ssh package (paramiko) can return similar file handler. We may consider get code more consistent. Please feel free to take this task.
`Improve local and remote process, so it can async/sync run, continual output. Try pexpect. `